[MIG]custom_attribute - Migration to 12.0 and refactor

clementmbr opened the pull request:
[WORK IN PROGRESS] Do not merge.

I made the first simple modifications to resolve a basic migration to 12.0 of `base_custom_attribute` and `product_custom_attribute`.
Then I made a refactor to change the way `attribute.set`, `attribute.group` and `attribute.attribute` are linked together. Moving from :
```
SET > o2m > GROUPS > o2m > ATTRIBUTES
```
to
```
SET > m2m > ATTRIBUTES > m2o > GROUP
```
For the moment the modules might be operational without blocking bugs.
The **work in progress** aim to use this simplified structure to display the attributes directly in the product form view instead of building them each time via the `product_attributes_form_view` wizard.

Cc @sebastienbeau @rvalyi

clementmbr committed:
[MIG] base_custom_attribute: Migration to 12.0

clementmbr committed:
[FIX] unidecode e base_custom_attribute groups

clementmbr committed:
[FIX] context eval in action views

clementmbr committed:
[REF] New structure relations between sets, groups and attributes

sebastienbeau commented on base_custom_attribute/models/attribute_attribute.py:
```diff
@@ -86,25 +82,31 @@ def _build_attribute_field(self, page, attribute):
 
     @api.model
-    def _build_attributes_notebook(self, attribute_group_ids):
+    def _build_attributes_notebook(self, attribute_set_id):
```
You need to change the logic here.
Now we do not care about the attribute_set_id selected (remove it).
What you are going to do is to build a view with all the attribute you have (that are used by a attribute_set)
Then for each field do something like

<field name="x_my_custom_field" attrs="{'invisible': [('attribute_set_id', 'not in', attribute_set_ids)])>

And the list of attributes set ids are my_custom_field.attribute_set_ids.ids

So every  field will be in the view but only the field used by an attribute set will be display.

clementmbr replied:
I tried to do that now, let me know what you think about it!

clementmbr committed:
[ADD] Display Attributes directly in the product view form

sebastienbeau commented on product_custom_attribute/views/product.xml:
```diff
@@ -1,23 +1,33 @@
             <field name="arch" type="xml">
                 <field name="type" position="after">
-                    <field name="attribute_set_id" context="{'default_model_id': %(product.model_product_template)d}"/>
+                    <field name="attribute_set_id"
```
can you move this field into the tab of attribute. It's will be improve the UI

clementmbr replied:
Thinking of it, don't you think it would be even better to put this attribute_set field under the Product's name, to follow the position of the `parent_id`field in Contacts form view :
![image](https://user-images.githubusercontent.com/31664455/77775088-d43e2000-702a-11ea-9abf-cbd7a1084575.png)

sebastienbeau commented on base_custom_attribute/models/attribute_attribute.py:
```diff
@@ -36,75 +35,83 @@ class AttributeAttribute(models.Model):
 
     @api.model
-    def _build_attributes_notebook(self, attribute_group_ids):
+    def _build_attributes_notebook(self, attribute_ids):
```
After thinking twice maybe it will be better to drop this notebook.
Most of customer to not have a lot of attributes
Maybe it's better to simply build group (with a string) per attribute group.
So instead of having the attribute in multiple page, we have one page but with several separator

Technique
field_1                             field_2
field_3                             field_4

Description
field_5                            field_6

....

sebastienbeau commented:
also can you improve the UI for text field (no label flag and col=4) and adding the label before

clementmbr commented:
> also can you improve the UI for text field (no label flag and col=4) and adding the label before

I didn't understand @sebastienbeau , you think it would be better to display a text in the full screen width (colspan = 4) ? 

I think it looks easier to read if the text is concentrated on the half of the screen (colspan = 2) :
![image](https://user-images.githubusercontent.com/31664455/77773694-bcfe3300-7028-11ea-8b07-728473b8b55d.png)
Than when it is on the full width (colspan = 4):
![image](https://user-images.githubusercontent.com/31664455/77773842-ee76fe80-7028-11ea-8de3-ca2e7ed505f2.png)
... and even more for shorter text.

Anyway, if you confirm you prefer to display Text fields on full width, let me know it and i'll fix it.

sebastienbeau commented:
If you have a mix of text field and non text field it appear like this

![image](https://user-images.githubusercontent.com/1164578/77777676-2e9da780-7050-11ea-91ed-0f8ce4b121d9.png)

Having a col=4 will force to push the other field after or before.

But indeed on big screen  col=2 is more readable (also we all love pep8 ;) )

So I think we have two possibility
- having col=4 (easy to do)
- adding separator before/after and keep col=2

Note: Odoo always put col=4 for text field so I prefer to be consistent.
for improving UI we can have a module that reduce the width of all text field to a max value

clementmbr commented:
> So I think we have two possibility
> 
> * having col=4 (easy to do)
> * adding separator before/after and keep col=2

So I tried the first option (col=4) :
![image](https://user-images.githubusercontent.com/31664455/77812182-69b6cf80-707e-11ea-8f8d-3caf704d51bd.png)

Then I tried to add the `<nextline/>` element to keep everything in half of the screen :
![image](https://user-images.githubusercontent.com/31664455/77812199-8b17bb80-707e-11ea-8b5b-cc52651b7225.png)

But I think I prefer a **third option** I tought about, displaying **each group in half of the screen** :
![image](https://user-images.githubusercontent.com/31664455/77812231-bdc1b400-707e-11ea-88d3-37dcafdf165b.png)
In case of more than 2 groups there might be a blank space between the last field of one group and the group two times after... But it looks ok to me, make some sense in a way...
(it's not a problem for small screens as the second group go under the first one if the screen is too small).

What do you think about it ? @sebastienbeau 
It's not col=4 as usual, but we can think that "usually" text fields in Odoo are used at the end of the screen (like Internal Notes) and never "need" to be displayed next to other fields or texts...
As the objective is different, it can justify our "inconsistency" ;)

clementmbr committed:
[ADD] Improve UI

bealdav commented:
Nice third option

clementmbr committed:
[FIX] Resolve Relational Model bug

hparfr commented:
Functional test:

Configuration > Product Attribute Sets > Create
>Set a name, add a line
>Create an attribute
>Set field label, type, 
> Create attribute Group
> Save attribute group
> Save attribute

Error:
` The operation cannot be completed:
- Create/update: a mandatory field is not set.
- Delete: another model requires the record being deleted. If possible, archive it instead.

Model: Attribute Set (attribute.set), Field: Name (name) 
`

hparfr commented:
I get a lot of warnings : 
`db odoo.addons.product.models.product: A matching product.template.attribute.value was not found for the product.attribute.value #13 on the template #30 
`

sebastienbeau commented:
@clementmbr last option seem to be the best visualy can you check the issue of Raph?

hparfr commented:
I had this issue i can't reproduce anymore: sometimes the attribute tab is empty when comming from sale > products > products. When coming from sale > products  > open product by attribute set it works.
After a reload of the page the error is gone. I suspect that in certains circonstances it get's badly loaded . I can't reproduce it yet.

hparfr commented:
Another issue: when you set an attribute with multi select, I get  odoo.exceptions.CacheMiss: ('ir.model.fields(11024,).name', None) and RecursionError: maximum recursion depth exceeded

clementmbr committed:
[FIX] Bug creating an Attribute from a creation of attribute.set and improve attribute form view layout

clementmbr commented:
> Functional test:
> 
> Configuration > Product Attribute Sets > Create
> 
> > Set a name, add a line
> > Create an attribute
> > Set field label, type,
> > Create attribute Group
> > Save attribute group
> > Save attribute
> 

@hparfr I might have resolved this bug. It was because the `default_attribute_set_id` passed to the new attribute didn't exist while we where creating the attribute.set. I've added a button step to save the new attribute.set before adding attributes :

![image](https://user-images.githubusercontent.com/31664455/78415289-2f59af00-75f7-11ea-95ba-f72ecd31a931.png)

----

> I get a lot of warnings :
> `db odoo.addons.product.models.product: A matching product.template.attribute.value was not found for the product.attribute.value #13 on the template #30 `

About this one I really don't have any idea to resolve it. It looks very strange as I don't know any object with this king of `product.attribute.value` or `product.template.attribute.value`....
The attributes are named `attribute.attribute`, they are linked to `product.template` through their field `attribute.attribute.model_id` and once the attribute becomes a product.template's field... it's a field with its field's name.

----


> I had this issue i can't reproduce anymore: sometimes the attribute tab is empty when comming from sale > products > products. When coming from sale > products > open product by attribute set it works.

I tried on my side in different ways but it never happened, so tell if you manage to reproduce the bug...

----


> Another issue: when you set an attribute with multi select, I get odoo.exceptions.CacheMiss: ('ir.model.fields(11024,).name', None) and RecursionError: maximum recursion depth exceeded

When does this problem occurs exactly ? Here I can create multiselect attributes without any problem... But I only have this kind of "RecursionError" when I launch the Odoo shell after closing my session. 
It's very strange... If you have any suggestion about where I can investigate to understand where does it come from, I'll be grateful !

Anyway, Merci beaucoup !

hparfr commented:
Hi, 
First issue during create is now solved. Thanks.

About, the issue of empty view, I have it often. a variant of it (all within the same brother tab): 
1) create a select field, bind it to contacts model. Add some records.
2) Then go to a product, you will be able to bind.
3) Go back to the attribute, add other records.
4) Then go to a product, you will not be able to bind to new records until you reload the page.

Btw it seems a delete constraint / cascade is missing:

5) Now delete a partner binded to a product (through custom attribute).
6) go back to the product page, you will get an error
` File "/odoo/links/base_custom_attribute/models/attribute_attribute.py", line 69, in <listcomp>
    ids = [op.value_ref.id for op in self.option_ids]
AttributeError: 'bool' object has no attribute 'id'
`

hparfr commented:
A second issue with select and a relation model:  you can't remove any record.

1) Create a field: select, relationnal model: contact.
2) Load attribute options, pick one contact
3) save
4) edit
5) on attribute options, click on the trash icon.
6) save.

You get : 
```py
  File "/odoo/links/product_custom_attribute/models/attribute_attribute.py", line 63, in write
    product.write(
  File "/odoo/links/product_custom_attribute/models/product.py", line 27, in write
    super(ProductTemplate, self).write(vals)
  [...]
  File "/odoo/src/odoo/sql_db.py", line 225, in execute
    res = self._obj.execute(query, params)
psycopg2.ProgrammingError: column "x_ffffffff" is of type integer but expression is of type record[]
LINE 1: UPDATE "product_template" SET "x_ffffffff"=ARRAY[(3, 14, 0)]... 
`

hparfr commented:
The bug about the multiselect:

(on a new fresh db).
1) Create a product attribute.
Type: multiselect,
relational model: Attribute value,
2) Load attribute options: select all.
3) Save.

I get : 
```py
 odoo.addons.base.models.ir_model: Failed to load field product.template.x_mul: skipped 
Traceback (most recent call last):
  File "/odoo/src/odoo/api.py", line 1049, in get
    value = self._data[key][field][record._ids[0]]
KeyError: 4743

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/odoo/src/odoo/fields.py", line 1005, in __get__
    value = record.env.cache.get(record, self)
  File "/odoo/src/odoo/api.py", line 1051, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: ('ir.model.fields(4743,).name', None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/odoo/src/odoo/addons/base/models/ir_model.py", line 977, in _add_manual_fields
    field = self._instanciate(field_data)
  File "/odoo/src/odoo/addons/base/models/ir_model.py", line 967, in _instanciate
    attrs = self._instanciate_attrs(field_data)
  File "/odoo/links/mail/models/ir_model.py", line 95, in _instanciate_attrs
    attrs = super(IrModelField, self)._instanciate_attrs(field_data)
  File "/odoo/links/base_sparse_field/models/models.py", line 63, in _instanciate_attrs
    attrs['sparse'] = serialization_record.name
  File "/odoo/src/odoo/fields.py", line 1009, in __get__
    self.determine_value(record)
  File "/odoo/src/odoo/fields.py", line 1112, in determine_value
    record._prefetch_field(self)
  File "/odoo/src/odoo/models.py", line 2860, in _prefetch_field
    result = records.read([f.name for f in fs], load='_classic_write')
  File "/odoo/src/odoo/models.py", line 2815, in read
    vals[name] = convert(record[name], record, use_name_get)
  File "/odoo/src/odoo/fields.py", line 2203, in convert_to_read
    return value.id
AttributeError: 'ir.model' object has no attribute 'id'

`

clementmbr committed:
[FIX] Rectify domain passed to add new Options to Multiselect Attributes

clementmbr committed:
[FIX] Bug when deleting an object pointed out by an attribute.option

clementmbr commented:
Thank you again @hparfr for your detailed reviews !

In this [last commit](https://github.com/shopinvader/odoo-pim/pull/6/commits/88e924c88e5aa82fdf94f4fa866083da674e9a01) I resolved the [first bug you mentioned](https://github.com/shopinvader/odoo-pim/pull/6#issuecomment-609765244) which occurs when deleting a Contact (or any object) linked to an attribute.option.

But the fact is that I didn't find a way to **properly delete the attribute.option** "when deleting the object pointed out by this option".

The problem is that an option and his related object _are not oficial related fields_, the option only as a `value_ref` field which is `Reference` type. (this is needed because we don't know _a priori_ the model of the related object pointed out by an option, until we create the attribute.attribute with his `relation_model_id` field).

So I don't see any possibility to do anything "when the related object pointed out by the option" is deleted", as we don't know his model.

What I did is only to add conditions to avoid dealing with options whose field `value_ref` is false, [like here](https://github.com/shopinvader/odoo-pim/pull/6/commits/88e924c88e5aa82fdf94f4fa866083da674e9a01#diff-0fcaef8880e4772a12151f7e9325c9dfR70) and [I made invisible the options that have no more `value_ref`](https://github.com/shopinvader/odoo-pim/pull/6/commits/88e924c88e5aa82fdf94f4fa866083da674e9a01#diff-b96b829393128531edb03bfbf82d8cc1R46) in the options tree (in the attribute view).

The only problem with that, is that an invisible line in a tree view is not "totally" invisible. There is still a weird empty line :
![image](https://user-images.githubusercontent.com/31664455/78590891-f39e2f80-7818-11ea-9389-f361a8e87f38.png)

So to avoid that I've thought about deleting the options with no `value_ref` but I didn't really know when to do that... The less worst option I've imagined was [on clicking the "Load Attribute Options" button](https://github.com/shopinvader/odoo-pim/pull/6/commits/88e924c88e5aa82fdf94f4fa866083da674e9a01#diff-0fcaef8880e4772a12151f7e9325c9dfR125), imagining that if the user delete an object that was once linked to an attribute option, he might want to add a new option soon...

But I agree that there might be a better place to clean these attribute.options... If anyone has a suggestion I'm opened to it !
Thanks !

-----

PS : When you said 
> 4.Then go to a product, you will not be able to bind to new records until you reload the page.

Is it a bug to be corrected ?
In fact when we add records to the options we change the datas in the database, so yes, these changes won't appear anywhere before we reload the page...
I don't see how it would be possible to improve that, but as always, if you have a suggestion, I'll like to hear it ! :)

hparfr commented on base_custom_attribute/models/attribute_attribute.py:
```diff
@@ -66,7 +66,11 @@ def _build_attribute_field(self, subgroup):
+                    for option in self.option_ids:
+                        # Display only options linked to an existing object
+                        if option.value_ref:
+                            ids.append(option.value_ref.id)
```
FYI, you can write it in a comprehension list as well.
```py
ids = [op.value_ref.id for op in self.option_ids if option.value_ref] 
```

clementmbr replied:
That's right, I even write it like that in another place. 
Made the correction.

clementmbr committed:
[FIX] Bug deleting option of Attribute Select with relation_model_id

clementmbr commented on product_custom_attribute/models/attribute_attribute.py:
```diff
@@ -66,7 +66,11 @@ def write(self, vals):
                             if option_id.value_ref:
-                                product.write(
-                                    {custom_field: [(3, option_id.value_ref.id, 0)]})
+                                if self.attribute_type == 'select':
```
@hparfr this might fix [the bug you mentioned here.](https://github.com/shopinvader/odoo-pim/pull/6#issuecomment-609770326)
I forgot to treat differently the many2one fields and the many2many, now it's ok.

hparfr replied:
yes it looks like, will test it tomorrow

hparfr commented on product_custom_attribute/models/attribute_attribute.py:
```diff
@@ -47,8 +47,15 @@ def _build_attributes_notebook(self, vals):
+        """ When updating the domain field or deleting an attribute.option delete
+        all the related attribute values in existing products"""
+
+        if ("domain" in list(vals.keys()) and self.relation_model_id and
```
vals.get('domain', False) instead of "domain" in list(vals.keys())

domain can't be None or "[]", but can it be [] or False ?

clementmbr replied:
You're right, domain can be '[]' or False (if it is modified), so as I want to delete the current options if a new domain "different from False or []" is registered, a better way to write my condition would be :
```
if vals.get('domain', False) not in [False, '[]'] and self.relation_model_id:
```
That's right ?

hparfr replied:
After looking closely I don't get it well what / why you are doing here.

About "
        """ When updating the domain field or deleting an attribute.option delete
        in existing products"""
        all the related attribute values in existing products""" 

Why ? If domain changes, you may not want to impact all products tied to the values. attributes can be the same with a different domain.

Then, why did you choose to not you use a @api.depends('domain', 'relation_model_id') ?
The removal of the option_ids can be done in unlink I guess.

clementmbr replied:
As we talked yesterday, [this commit](https://github.com/shopinvader/odoo-pim/commit/dd77b5b53feb6ffa1df54cb61b7e8d1c356722a7) might resolve the question and delete the existing option values in product ONLY if this value is not in the new domain/options available.

hparfr commented:
>    4.Then go to a product, you will not be able to bind to new records until you reload the page.
> Is it a bug to be corrected ?

Yes, it's like the empty page bug. Did you try to reproduce it ?
When you navigate from a menu to another, data are reloaded from the database and the page is not reloaded, the views are still in cache only the data is loaded (search_read, read, name_get...), . Here it didn't work until I reload the page from the browser (new request to load_views, etc.).

clementmbr committed:
[IMP] raise error in onchange domain and improve syntax

clementmbr commented:
> The bug about the multiselect:
> 
> (on a new fresh db).
> 
> 1. Create a product attribute.
>    Type: multiselect,
>    relational model: Attribute value,
> 2. Load attribute options: select all.
> 3. Save.

About this bug @hparfr I didn't manage to reproduce it. I created as you said a new product.attribute type multiselect, I linked it do the model Attribute Value and I managed to save it and to use is in products as other normal attributes... Strange... Maybe the first commits I made today fixed it without noticing it...
Let me know it still occur on your side !
Merci ! :sparkles:

clementmbr committed:
[FIX] On editing Attribute, delete product attribute values only if they are not in the new domain/options

clementmbr committed:
[IMP] Default and onchange fill Product's attribute_set with his Category's attribute_set

clementmbr committed:
[IMP] Delete Attribute's related Field when deleting Attribute

clementmbr committed:
[FIX] Prevent from changing 'JSON Field'

clementmbr committed:
[FIX] Allow Multiselect Attributes to be serialized or not

clementmbr committed:
[IMP] Improve UI navigation with domain and context

hparfr commented:
Hi @clementmbr.

Thanks, I have tested multiselect  without json and it works way better. Vue issues, cache miss and infinite recursion do not appear anymore. 

I have found two issues.

Issue 1 : more than one multiselect

1) create an attribute multiselect, without json, without relational model, add some values.
2) go to a product, choose some values
3) create a second attribute with multi select, without json, without relational model, add some other values.
4) go back to the product, you will see two fields, but all the values will be the same (of the first one). 
 

Issue 2: behavioral difference for attribute deletion
1) Create a multi select with a relational model, add some values.
2) Create a multi select without a relational model, add some values
3) Go to a product, set some values for both fields
4) Go back to the fields, remove some values
5) On the product, you will see: 
- values from the relational model will be kept
- values from the other field will be removed

clementmbr commented:
> I have found two issues.
> 
> Issue 1 : more than one multiselect
> 
> 1. create an attribute multiselect, without json, without relational model, add some values.
> 2. go to a product, choose some values
> 3. create a second attribute with multi select, without json, without relational model, add some other values.
> 4. go back to the product, you will see two fields, but all the values will be the same (of the first one).
> 
@hparfr Issue 1 solved with [this latest commit.](https://github.com/shopinvader/odoo-pim/commit/57a99ec4febc4e7f53248cc461a467b5e727832d)

hparfr commented:
> @hparfr Issue 1 solved with [this latest commit.](https://github.com/shopinvader/odoo-pim/commit/57a99ec4febc4e7f53248cc461a467b5e727832d)

Thanks it works as expected.

clementmbr committed:
[FIX] Avoid relation_table's name conflict with MultiSelect Attributes non-serialized

clementmbr commented:
> 
> Issue 2: behavioral difference for attribute deletion
> 
> 1. Create a multi select with a relational model, add some values.
> 2. Create a multi select without a relational model, add some values
> 3. Go to a product, set some values for both fields
> 4. Go back to the fields, remove some values
> 5. On the product, you will see:
> 
> * values from the relational model will be kept
> * values from the other field will be removed

@hparfr Issue 2 was born because of a mistake I made trying to improve the code... I've just corrected it in the last update of the last commit :
https://github.com/shopinvader/odoo-pim/commit/9b2883d68210d9e0206af25788e117b3df726c55

sebastienbeau committed:
[IMP] simplify UI

Only create group in tree edit
Simplify code create attribute from attribute_set

sebastienbeau committed:
[REF] remove newline and use bootstrap class

clementmbr committed:
[IMP] Attribute Set required condition and Sequences names

sebastienbeau commented on base_custom_attribute/views/attribute_attribute_view.xml:
```diff
@@ -1,36 +1,60 @@
+                            widget="many2many_tags"
+                            domain="[('model_id', '=', context.get('default_model_id'))]"
+                            options="{'no_create': 1}"
+                            required="not context.get('from_attribute_set')"
```
@hpar @clementmbr I do not know if it's a good idea to make this field required?
Indeed maybe you want to create your attribute (all of them and after create your attribute set)?

clementmbr replied:
I don't really manage to imagine this use case to create first the attributes and then the attribute set... But if we what to cover this kind of scenario, we might add some "Action button" on attributes tree view to give an attribute set to many attributes at once, might we?

In my opinion, I think I like this field to be required as it will help new users to understand what are the necessary fields to use an attribute (as it is impossible to use an attribute without an attribute set, a new user can create an attribute without attribute set at first and then loosing some time to understand why his attribute doesn't appear anywhere...)

hparfr replied:
> I don't really manage to imagine this use case to create first the attributes and then the attribute set.

For me it make sense to first create the attributes in a row, then figure out how to group them.

sebastienbeau committed:
[FIX] Update precommit config and cleanup the code

sebastienbeau commented:
@clementmbr I have cleanned the pep8 issue.
Before adding new commit, please run "pre-commit install" in the repository this will install all the hook needed for automagically cleanning the code

sebastienbeau committed:
[FIX] fix test

sebastienbeau commented:
@clementmbr \o/ test are green, let's add more, I am on it

sebastienbeau committed:
[REF] remove useless code

sebastienbeau commented on product_custom_attribute/models/attribute_attribute.py:
```diff
@@ -0,0 +1,82 @@
+from odoo.osv.orm import setup_modifiers
+
+
+class AttributeAttribute(models.Model):
```
We should move this code into the base module

bealdav commented:
Just test it works, thanks a lot.

A small problem however, 

The screen on product.product don't let appears attributes created on products

![2020-04-15_11-07](https://user-images.githubusercontent.com/1853434/79320174-a7827780-7f09-11ea-9597-1ebad9d5c7ef.png)

while it appears normally in product.template

![2020-04-15_11-06](https://user-images.githubusercontent.com/1853434/79320184-abae9500-7f09-11ea-9a84-a83e91359ca0.png)

Maybe I misunderstood something ?

hparfr commented:
> Just test it works, thanks a lot.
> 
> A small problem however,

It's a known bug : you have to reload the whole page after adding some fields;

bealdav commented:
In fact even I don't add fields, I just apply attribute set on product.product, and reload don't works. But may be a bad practice of me

bealdav commented:
It seems it's because I can't create an attribute set != product.template.

Maybe it's normal

hparfr commented:
I have a bug with the translation.
After a fresh install of product_custom_attribute, when I load the  translation for french with Overwrite Existing Terms.
Then I get: 

>psycopg2.ProgrammingError: ON CONFLICT DO UPDATE command cannot affect row a second time
HINT:  Ensure that no rows proposed for insertion within the same command have duplicate constrained values.

After deleting both fr.po from base_custom_attribute and product_custom_attribute it works.

clementmbr commented:
> I have a bug with the translation.
> After a fresh install of product_custom_attribute, when I load the translation for french with Overwrite Existing Terms.
> Then I get:
> 
> > psycopg2.ProgrammingError: ON CONFLICT DO UPDATE command cannot affect row a second time
> > HINT:  Ensure that no rows proposed for insertion within the same command have duplicate constrained values.
> 
> After deleting both fr.po from base_custom_attribute and product_custom_attribute it works.

@hparfr  I simply deleted the .po files in my last commit to avoid these bugs. Better have no translation at all than a bad translation on old fields names...

clementmbr committed:
[REF] No more atttribute.attribute methods in product_custom_attribute

clementmbr committed:
[FIX] Delete .po files before doing a clean translation

sebastienbeau committed:
[IMP] add test for testing rendering of view

sebastienbeau committed:
[REF] start to refactor in order to have green test

sebastienbeau committed:
[IMP] improve UI for sorting attribute

sebastienbeau committed:
Merge pull request #2 from akretion/12.0-clean-and-test

[REF] remove useless code

hparfr committed:
Add description

clementmbr committed:
[FIX] rename _propagate_new_options() and object variable

'object' is a reserved builtin name in python...

clementmbr committed:
[FIX] test_build_view OK

clementmbr committed:
[ADD] pim_base module

clementmbr committed:
[REF] rename module name 'product_custom_attribute' to 'pim_custom_attribute'

clementmbr committed:
[REF] move pim_custom_attribute in the PIM menu

clementmbr committed:
[REF] remove UoM menu in pim_base

clementmbr committed:
[IMP] create attribute related to existing fields

Add an option "Native" or "Custom" when creating a new
'attribute.attribute' to allow relating it to an existing
'ir.model.fields' field or a new "manual" one.

clementmbr committed:
[REF] Add Attr related to native fields in Attr view

...depending on context "include_native_attribute".

Commit include tests and new "Native Attributes" demo records in
pim_custom_attribute module.
It also moves the code overiding the model's fields_view_get from
"product.template" to the "attribute.set.owner.mixin".

By doing so, the model using base_custom_attribute only need to :

- _inherit the mixin "attribute.set.owner.mixin"
- add a placeholder `<separator name="attributes_placeholder" />` at
the desired location in the model's form view.
- add a context `{"include_native_attribute": True}` on the action
leading to this form view if it is needed to display attributes related
to native fields together with the other "custom" attributes.

clementmbr committed:
[REF] rename modules based on attribute_set and pim

New organization :

- attribute_set (former base_custom_attribute)
- product_attribute_set (former pim_custom_attribute but without menus)
- pim (The "PIM application" former pim_base)
- pim_attrubute_set (depends on product_attribute_set and adds menus in
the PIM application)

clementmbr committed:
[FIX] when including native field not present in form

clementmbr committed:
[REF] rename 'attribute_nature' field into 'nature'

And small FIX in _build_attribute_field()

clementmbr committed:
[IMP] Add REAME to the 4 modules

clementmbr committed:
[FIX] propagate readonly attribute for included native attributes

clementmbr committed:
[IMP] base.user_admin in Attribute Set and PIM Manager group

clementmbr committed:
[IMP] Move Product's attributes_placeholder to product_attribute_set

clementmbr committed:
[IMP] remove class in super() and other misc improvements

clementmbr committed:
[FIX] remove tests in product_attribute_set

clementmbr committed:
[IMP] remove attribute_set access rights for group_no_one

clementmbr committed:
[REF] remove unecessary module name in External ID

clementmbr committed:
[REF] remove @api.multi

clementmbr committed:
[IMP] Add test write on partner's attributes

clementmbr committed:
[IMP] Rename "Advanced Attribute Set settings" group

sebastienbeau committed:
Merge pull request #3 from akretion/12.0-pim-ref-modules-names

[12.0][REF] PIM: new module pim_base and upgrade base_custom_attribute

sebastienbeau commented:
/ocabot merge nobump

sebastienbeau commented:
/ocabot merge nobump
